### PR TITLE
Add output scaling filtering options

### PR DESCRIPTION
**Pull request by xperia64**
Partially addresses https://github.com/RPCS3/rpcs3/issues/11187 by adding a bilinear/nearest option and provides a config/UI implementation so that more scaling/filtering options may be added later.

As per @kd-11, FSR is currently unimplemented for OpenGL but is intended to be added later, so for now, while FSR can be selected under OpenGL, it just falls to bilinear.

The new combobox option is shown here:
image

The options are Nearest, Bilinear (default) and FSR.

### CODE REVIEW

File: rpcs3/rpcs3qt/settings_dialog.cpp [hunk 1]
```diff
@@ -599,6 +599,18 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> gui_settings, std
...
+	SubscribeTooltip(ui->outputScalingMode, tooltips.settings.output_scaling_mode);
+	connect(ui->outputScalingMode, QOverload<int>::of(&QComboBox::currentIndexChanged), this, [r_creator, this](int index)
+	{
+		const bool is_vulkan = (ui->renderBox->currentText() == r_creator->Vulkan.name);
+		const QVariantList var_list = ui->outputScalingMode->itemData(ui->outputScalingMode->currentIndex()).toList();
+		ensure(var_list.size() == 2 && var_list[1].canConvert<int>());
+		const bool fsr_selected = static_cast<output_scaling_mode>(var_list[1].toInt()) == output_scaling_mode::fsr;
+		ui->fsrSharpeningStrength->setEnabled(is_vulkan && fsr_selected);
+		ui->fsrSharpeningStrengthReset->setEnabled(is_vulkan && fsr_selected);
+	});
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/settings_dialog.cpp:612]: I'd rather have one lambda that is triggered whenever we change either settings.

```
const auto enableFsrSettings = [this]()
{
	const bool is_vulkan = (ui->renderBox->currentText() == r_creator->Vulkan.name);
	const QVariantList var_list = ui->outputScalingMode->itemData(ui->outputScalingMode->currentIndex()).toList();
	ensure(var_list.size() == 2 && var_list[1].canConvert<int>());
	const bool fsr_selected = static_cast<output_scaling_mode>(var_list[1].toInt()) == output_scaling_mode::fsr;
	ui->fsrSharpeningStrength->setEnabled(is_vulkan && fsr_selected);
	ui->fsrSharpeningStrengthReset->setEnabled(is_vulkan && fsr_selected);
};
```
`connect(ui->outputScalingMode, QOverload<int>::of(&QComboBox::currentIndexChanged), this, enableFsrSettings);`

File: rpcs3/rpcs3qt/settings_dialog.cpp [hunk 2]
```diff
@@ -866,9 +875,12 @@ settings_dialog::settings_dialog(std::shared_ptr<gui_settings> gui_settings, std
...
 		ui->vulkansched->setEnabled(is_vulkan);
-		ui->fsrUpscalingEnable->setEnabled(is_vulkan);
-		ui->fsrSharpeningStrength->setEnabled(is_vulkan);
-		ui->fsrSharpeningStrengthReset->setEnabled(is_vulkan);
+
+		const QVariantList var_list = ui->outputScalingMode->itemData(ui->outputScalingMode->currentIndex()).toList();
+		ensure(var_list.size() == 2 && var_list[1].canConvert<int>());
+		const bool fsr_selected = static_cast<output_scaling_mode>(var_list[1].toInt()) == output_scaling_mode::fsr;
+		ui->fsrSharpeningStrength->setEnabled(is_vulkan && fsr_selected);
+		ui->fsrSharpeningStrengthReset->setEnabled(is_vulkan && fsr_selected);
```
Comment (Megamouse) [inline rpcs3/rpcs3qt/settings_dialog.cpp:883]: then remove this and add an additional connect:
`connect(ui->outputScalingMode, QOverload::of(&QComboBox::currentIndexChanged), this, enableFsrSettings);`
and call
`enableFsrSettings();`
once in the end to ensure they are enabled/disabled

File: rpcs3/Emu/RSX/GL/GLOverlays.cpp [hunk 3]
```diff
@@ -603,11 +603,17 @@ namespace gl
...
-		input_filter = GL_LINEAR;
+		m_input_filter = gl::filter::linear;
 	}
 
-	void video_out_calibration_pass::run(gl::command_context& cmd, const areau& viewport, const rsx::simple_array<GLuint>& source, f32 gamma, bool limited_rgb, bool _3d)
+	void video_out_calibration_pass::run(gl::command_context& cmd, const areau& viewport, const rsx::simple_array<GLuint>& source, f32 gamma, bool limited_rgb, bool _3d, gl::filter input_filter)
 	{
+		if (m_input_filter != input_filter)
+		{
+			m_input_filter = input_filter;
```
Comment (Megamouse) [inline rpcs3/Emu/RSX/GL/GLOverlays.cpp:613]: I think the member should stay a GLenum and then you should use a switch at the beginning of the function to set the enum properly.

Then you also won't have do those static casts everywhere, which fit by coincidence at best.

Comment (xperia64) [reply rpcs3/Emu/RSX/GL/GLOverlays.cpp:613]: I felt the enum was more explicit here and based the static casting off of https://github.com/RPCS3/rpcs3/blob/master/rpcs3/Emu/RSX/GL/glutils/fbo.cpp#L29

File: rpcs3/rpcs3qt/settings_dialog.cpp (hunk 1 above)
Comment (xperia64) [reply rpcs3/rpcs3qt/settings_dialog.cpp:612]: Will do.

File: rpcs3/Emu/RSX/GL/GLOverlays.cpp (hunk 3 above)
Comment (kd-11) [reply rpcs3/Emu/RSX/GL/GLOverlays.cpp:613]: For the record, a lot of the gl:: namespace enums are rather pointless and I'm not the one who came up with them. They only make sense for the purposes of type checking, i.e making sure the correct type of enum is passed to a function. But in this case its being used for storage, there is really no reason to keep it an enum.
The only benefit I can think of is that msvc is really bad at inferring GL_XXX enums during debug sessions.

Comment (kd-11) [reply rpcs3/Emu/RSX/GL/GLOverlays.cpp:613]: TLDR, it's fine.

**Review by kd-11 — APPROVED** [review]
GL/VK side looks ok. I'll leave @Megamouse to review Qt/UI side.

File: rpcs3/Emu/RSX/GL/GLOverlays.cpp (hunk 3 above)
Comment (Megamouse) [reply rpcs3/Emu/RSX/GL/GLOverlays.cpp:613]: Ah. for some reason I though the parameter was a config param when I read this at 1am xD

**Review by Megamouse — CHANGES_REQUESTED** [review]
See comments

**Comment by 4tl4sh** [comment]
> "Popping out" options based on other options is a horrible choice and will make the UI jumpy.

I can't agree. I was suggesting that as a UX designer and as a user. If, for any case, every scaling option had its own settings, would you just show every option one after the other, disable them and call it a day?

You're not hiding features, you're hiding a sub-feature that belong to a choice that you have to make first: FSR. If no FSR is chosen, why would I clutter the screen with an option that isn't needed? The user could think: why is this option grayed out? Their only way to know how to gray out this option would be cycling between Output Scaling options mindlessly just to see what happens.

Jumpiness isn't beautiful, but having a cluttered UI where you don't know what does what unless you touch all the buttons is really bad idea. If B is a consequence of A, I shouldn't show you B first unless you want to play cluedo. An is basic that an interface is as self-explanatory and guided as possible, so cluedos are bad.

Sorry if the tone sounds bad, not my intention (wanted to clarify after reading my answer). I'm just giving you some depth and examples.

File: rpcs3/rpcs3qt/settings_dialog.cpp (hunk 1 above)
Comment (xperia64) [reply rpcs3/rpcs3qt/settings_dialog.cpp]: I have added an `apply_fsr_specific_options` lambda with the above under the similar `apply_renderer_specific_options` lambda and connected the two in the appropriate spot, removing the duplicate code.

**Comment by xperia64** [comment]
> > "Popping out" options based on other options is a horrible choice and will make the UI jumpy.
> 
> I can't agree. I was suggesting that as a UX/UI designer and as a user. If, for any case, every scaling option had its own settings, would you just show every option one after the other, disable them and call it a day?
> 
> You're not hiding features, you're hiding a sub-feature that belongs to a choice that you have to make first: FSR. If no FSR is chosen, why would I clutter the screen with a conditional setting that isn't needed? The user could think: why is this option grayed out? Their only way to know how to enable this option would be cycling between Output Scaling options mindlessly just to see what happens.
> 
> Jumpiness isn't beautiful, but having a cluttered UI where you don't know what does what unless you touch all the buttons is really bad idea. If B is a consequence of A, I shouldn't show you B first unless you want to play cluedo. It's a must that an interface is as self-explanatory and guided as possible and decisions like this are against this principle.
> 
> Sorry if the tone sounds bad, not my intention (wanted to clarify after reading my answer). I'm just giving you some depth and examples.
> 
> If UI consistency is a top priority for the product and it's mandatory to keep this conditional setting visible, maybe a hint can be added: "Chose FSR scaling to enable this setting". Description bottom box perphaps?

At the moment disabling settings when unusable is the standard in this UI, and this PR is not the place for more drastic UI change discussions.

There is already a tooltip hint explaining all of the options when highlighting over the filtering combobox.

**Comment by 4tl4sh** [comment]
> > > "Popping out" options based on other options is a horrible choice and will make the UI jumpy.
> > 
> > 
> > I can't agree. I was suggesting that as a UX/UI designer and as a user. If, for any case, every scaling option had its own settings, would you just show every option one after the other, disable them and call it a day?
> > You're not hiding features, you're hiding a sub-feature that belongs to a choice that you have to make first: FSR. If no FSR is chosen, why would I clutter the screen with a conditional setting that isn't needed? The user could think: why is this option grayed out? Their only way to know how to enable this option would be cycling between Output Scaling options mindlessly just to see what happens.
> > Jumpiness isn't beautiful, but having a cluttered UI where you don't know what does what unless you touch all the buttons is really bad idea. If B is a consequence of A, I shouldn't show you B first unless you want to play cluedo. It's a must that an interface is as self-explanatory and guided as possible and decisions like this are against this principle.
> > Sorry if the tone sounds bad, not my intention (wanted to clarify after reading my answer). I'm just giving you some depth and examples.
> > If UI consistency is a top priority for the product and it's mandatory to keep this conditional setting visible, maybe a hint can be added: "Chose FSR scaling to enable this setting". Description bottom box perphaps?
> 
> At the moment disabling settings when unusable is the standard in this UI, and this PR is not the place for more drastic UI change discussions.
> 
> There is already a tooltip hint explaining all of the options when highlighting over the filtering combobox.

Then I suggest to add a conditional tooltip hint that appears only when you mouse hover the disabled FSR Sharpening setting that invites to user to Select FSR Scaling to enable the setting. My final suggestion regarding this topic.

**Review by Megamouse — APPROVED** [review]
OK

**Pull request merged**